### PR TITLE
libbpf v0.8.1 (new formula)

### PR DESCRIPTION
**Pull request by rlankfo**
Signed-off-by: Robbie Lankford <robert.lankford@grafana.com>

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/libbpf.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+  def install
+    chdir "src" do
+      system "make"
+      system "make", "install", "PREFIX=#{prefix}"
+    end
+  end
+
+  test do
+    system "true"
```
Comment (rlankfo) [inline Formula/libbpf.rb:21]: TODO: Add a proper test. I'll check and see what other libs do.

**Comment by rlankfo** [comment]
I'm not sure why the test keeps failing. My guess is an issue with gcc-5? It works for me locally on bullseye, using a newer version of gcc:


```
$ uname -a
Linux bullseye 5.10.0-16-amd64 #1 SMP Debian 5.10.127-1 (2022-06-30) x86_64 GNU/Linux
```

```
$ brew config
HOMEBREW_VERSION: 3.5.5-23-g9c08c5b
ORIGIN: https://github.com/Homebrew/brew
HEAD: 9c08c5b8bac6a9795a8856007f942e32f5007e69
Last commit: 8 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: cc4cbc11eec2148fddb741a2b97791b5dcf73a61
Core tap last commit: 3 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: vim
HOMEBREW_MAKE_JOBS: 2
Homebrew Ruby: 2.6.8 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: dual-core 64-bit icelake
Clang: 14.0.6
Git: 2.30.2 => /bin/git
Curl: 7.74.0 => /bin/curl
Kernel: Linux 5.10.0-16-amd64 x86_64 GNU/Linux
OS: Debian GNU/Linux 11 (bullseye) (bullseye)
Host glibc: 2.31
/usr/bin/gcc: 10.2.1
/usr/bin/ruby: N/A
glibc: N/A
gcc@5: N/A
xorg: N/A
```

```
$ brew test --verbose libbpf
==> Testing libbpf
==> /home/linuxbrew/.linuxbrew/bin/gcc-11 test.c -I/home/linuxbrew/.linuxbrew/Cellar/libbpf/0.8.1/include -L/home/linuxbrew/.linuxbrew/Cellar/libbpf/0.8.1/lib -o test
==> ./test
0.8
```

**Comment by rlankfo** [comment]
> It's probably failing because it needs a newer version of the Linux headers: [libbpf/libbpf#53 (comment)](https://github.com/libbpf/libbpf/issues/53#issuecomment-580962821). First try adding `linux-headers@4.15` as a dependency and if that doesn't work try `linux-headers@5.16`. **EDIT** Since it looks like you already had to use `linux-headers@5.16` for `bcc`, you should use the same version here to maintain consistency.

Thanks @danielnachun 🙏  It seems this is the issue. Much appreciated.

File: Formula/libbpf.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
+class Libbpf < Formula
+  desc "Automated upstream mirror for libbpf stand-alone build"
+  homepage "https://github.com/libbpf/libbpf"
+  url "https://github.com/libbpf/libbpf/archive/refs/tags/v0.8.1.tar.gz"
+  sha256 "7bda8187efc619d1eb20a1ba5ab949dd68d40dd44945310c91ac0f915fa4a42b"
+  license "BSD-2-Clause"
+
+  depends_on "elfutils"
+  depends_on :linux
+  depends_on "linux-headers@5.16"
```
Comment (danielnachun) [inline Formula/libbpf.rb:10]:
```suggestion
  depends_on "linux-headers@5.16" => :test
```

File: Formula/libbpf.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      #include <bpf/bpf.h>
+      #include <stdio.h>
+
+      int main() {
+
+        printf("%d.%d", LIBBPF_MAJOR_VERSION, LIBBPF_MINOR_VERSION);
+        return(0);
+      }
+    EOS
+    system ENV.cc, "test.c", "-I#{include}", "-L#{lib}", "-lbpf",
```
Comment (danielnachun) [inline Formula/libbpf.rb:32]: This might also need `-I#{Formula["linux-headers@5.16"].opt_include}` as an include flag so that it can find the headers.

Comment (rlankfo) [reply Formula/libbpf.rb]: Thanks it looked like this worked. The test was only referencing macros before (so didn't need to link against the library), but I'm updating it to call a function which will require being linked.

File: Formula/libbpf.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  depends_on "linux-headers@5.16" => :test
+  depends_on "elfutils"
+  depends_on :linux
+  depends_on "pkg-config"
+  depends_on "zlib"
+
+  def install
+    chdir "src" do
+      system "make"
+      system "make", "install", "PREFIX=#{prefix}"
```
Comment (danielnachun) [inline Formula/libbpf.rb:17]:
```suggestion
      system "make", "install", "PREFIX=#{prefix}", "LIBSUBDIR=lib"
```
This should prevent the Makefile from installing to `lib64`, which is not allowed in Homebrew.  That will save you the trouble of trying to link to `lib64` in the test.

Comment (rlankfo) [reply Formula/libbpf.rb]: Thanks, I had just noticed this is well after investigating the libbpf Makefile a bit closer.

File: Formula/libbpf.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  homepage "https://github.com/libbpf/libbpf"
+  url "https://github.com/libbpf/libbpf/archive/refs/tags/v0.8.1.tar.gz"
+  sha256 "7bda8187efc619d1eb20a1ba5ab949dd68d40dd44945310c91ac0f915fa4a42b"
+  license "BSD-2-Clause"
+
+  depends_on "linux-headers@5.16" => :test
+  depends_on "elfutils"
+  depends_on :linux
+  depends_on "pkg-config"
+  depends_on "zlib"
```
Comment (danielnachun) [inline Formula/libbpf.rb:12]:
```suggestion
  uses_from_macos "zlib"
```
Sorry to nitpick but It's probably better for us to use this just to maintain consistency with other formulae, even though this is Linux-only so it looks a bit weird to reference macOS.  `uses_from_macos` lines come after regular `depends_on` lines with a blank line separating them for readability.

File: Formula/libbpf.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  homepage "https://github.com/libbpf/libbpf"
+  url "https://github.com/libbpf/libbpf/archive/refs/tags/v0.8.1.tar.gz"
+  sha256 "7bda8187efc619d1eb20a1ba5ab949dd68d40dd44945310c91ac0f915fa4a42b"
+  license "BSD-2-Clause"
+
+  depends_on "linux-headers@5.16" => :test
+  depends_on "elfutils"
+  depends_on :linux
+  depends_on "pkg-config"
+  uses_from_macos "zlib"
```
Comment (rlankfo) [inline Formula/libbpf.rb:12]:
```suggestion

  uses_from_macos "zlib"
```

**Review by carlocab — APPROVED** [review]
LGTM; we can merge this as soon as CI is green.

Thanks so much for your work, @rlankfo! You have no idea how thankful we are for contributors like you.

**Pull request closed without merging**